lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -0,0 +1,342 @@
+        self._charm.plugin_manager.run()
+
+
+...
+
+
+class OpenSearchBaseCharm(CharmBase):
+
+    def __init__(self, *args, distro: Type[OpenSearchDistribution] = None):
+        ...
+
+        self.my_plugin = MyPlugin(self)
+
+    ...
+
+    def _on_update_status(self, event):
+        ...
+
+        # Check my plugin status
+        if self.model.get_relation("my-plugin-relation") is not None:
+            self.unit.status = self.my_plugin.status()
+
+        for relation in self.model.relations.get(ClientRelationName, []):
+            self.opensearch_provider.update_endpoints(relation)
+
+        self.user_manager.remove_users_and_roles()
+        # If relation not broken - leave
+        if self.model.get_relation("certificates") is not None:
+            return
+        # handle when/if certificates are expired
+        self._check_certs_expiration(event)
+"""
+
+import logging
+from abc import abstractmethod
+from typing import Any, Dict, List, NamedTuple, Optional
+
+from charms.opensearch.v0.helper_enums import BaseStrEnum
+from charms.opensearch.v0.opensearch_exceptions import OpenSearchError
+from jproperties import Properties
+
+# The unique Charmhub library identifier, never change it
+LIBID = "3b05456c6e304680b4af8e20dae246a2"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+class OpenSearchPluginError(OpenSearchError):
+    """Exception thrown when an opensearch plugin is invalid."""
+
+
+class OpenSearchPluginMissingDepsError(OpenSearchPluginError):
+    """Exception thrown when an opensearch plugin misses installed dependencies."""
+
+
+class OpenSearchPluginInstallError(OpenSearchPluginError):
+    """Exception thrown when opensearch plugin installation fails."""
+
+
+class OpenSearchPluginRemoveError(OpenSearchPluginError):
+    """Exception thrown when opensearch plugin removal fails."""
+
+
+class OpenSearchPluginApplyConfigError(OpenSearchPluginError):
+    """Exception thrown when opensearch plugin applies a config (add or remove)."""
+
+
+class PluginState(BaseStrEnum):
+    """Enum for the states possible in plugins' lifecycle."""
+
+    MISSING = "missing"
+    INSTALLED = "installed"
+    ENABLED = "enabled"
+    WAITING_FOR_UPGRADE = "waiting-for-upgrade"
+
+
+class OpenSearchPluginConfig(NamedTuple):
+    """Represents configuration of a plugin to be applied in a given action.
+
+    An action can be: config() or disable().
+    Each returns a namedtuple containing the configurations and secrets to be added and
+    removed in that operation.
+    """
+
+    config_entries_to_add: Dict[str, str] = {}
+    config_entries_to_del: Dict[str, str] = {}
+    secret_entries_to_add: Dict[str, str] = {}
+    secret_entries_to_del: Dict[str, str] = {}
```

Comment:
We should perhaps use a regular class here
```suggestion
class OpenSearchPluginConfig:
    """Represent the configuration of a plugin to be applied when configuring or disabling it."""
    
    def __init__(
        config_entries_to_add: Optional[Dict[str, str]] = None,
        config_entries_to_del: Optional[Dict[str, str]] = None,
        secret_entries_to_add: Optional[Dict[str, str]] = None,
        secret_entries_to_del: Optional[Dict[str, str]] = None,
    )
        self.config_entries_to_add = config_entries_to_add or {}
        self.config_entries_to_del = config_entries_to_del or {}
        self.secret_entries_to_add = secret_entries_to_add or {}
        self.secret_entries_to_del = secret_entries_to_del or {}
```